### ISSUE TITLE
sql: simple bulk loading in Cockroach very slow (compared to PG)

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#5981]
It's slow to the point where you wonder how you're going to get data into it.
Did the following (on my MacBook).

```
go install github.com/tschottdorf/goplay/tblgen
rm -rf cockroach-data && cockroach start &
docker run -p 5432:5432 postgres

time (tblgen 10000 | cockroach sql)
```

> real    0m16.712s
> user    0m35.817s
> sys 0m5.158s

```
time (tblgen 10000 | psql -h $(docker-machine ip default) -U postgres)
```

> real    0m0.306s
> user    0m0.023s
> sys 0m0.024s

There ought to be some very low-hanging fruit there. `tblgen` has a constant that can tune how much we want to batch, but by default it inserts 640 entries per batch, which can't be extremely bad.

The typical trace (on the slower end) looks like this:

```
11:35:14.461839  .     3    ... node 1
11:35:14.461847  .     8    ... read has no clock uncertainty
11:35:14.462042  .   195    ... executing 2562 requests
11:35:14.462500  .   458    ... read-write path
11:35:14.462531  .    31    ... command queue
11:35:14.475034  . 12502    ... raft
11:35:14.480098  .  5064    ... applying batch
```

Some which need to renew the leader lease obviously take a little longer. I ran with a (horrible) hack that simply bypasses Raft, which didn't at all help:

> real  0m17.260s
> user    0m36.141s
> sys 0m5.156s

 But that's still slow. Might be time to push #5255 to completion for easier diagnosis here.

For testing sustained throughput, probably want

```
./cockroach zone set .default 'range_max_bytes: 99999999999'
./cockroach zone set .default 'replicas:
- attrs: []
'
```

for once a range splits, these inserts are not 1PC txns any more and things get really bad (on the other hand, background queues could get problematic with a very large replica).

**Comment by petermattis** [comment]
Fiddling with the number of inserted rows per statement I see:

```
# 640 rows per statement
real	0m14.717s
# 1 row per statement
real	0m8.752s
# 20 rows per statement
real	0m2.846s
```

**Comment by bdarnell** [comment]
Remember that in these traces the time deltas are "off" by a line: it spent 12ms in the command queue and 5ms in raft. Since smaller batches turned out to be cheaper, my guess is that large batches turn out to cover too much of the keyspace and are more likely to conflict with each other in the command queue.

**Comment by tschottdorf** [comment]
Interesting, that's unexpected since they're all point inserts. Maybe
there's something unexpected going on with the index updates. I'll take a
look tomorrow.

On Mon, Apr 11, 2016 at 5:14 PM Ben Darnell <notifications@github.com>
wrote:

> Remember that in these traces the time deltas are "off" by a line: it
> spent 12ms in the command queue and 5ms in raft. Since smaller batches
> turned out to be cheaper, my guess is that large batches turn out to cover
> too much of the keyspace and are more likely to conflict with each other in
> the command queue.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/5981#issuecomment-208564356>
>
-- 

-- Tobias

**Comment by petermattis** [comment]
Discovered something else strange. Change the random user id to use `i+j` instead and the timings change from:

```
real	0m12.526s
```

to:

```
real	0m5.226s
```

This isn't due to the removal of the `rand.Int63()` call (I kept the call around, but don't use the value). Something about inserting the rows in user-id order is a lot faster.

PS The numbers changed from my previous message due to the inclusion of #5966.

**Comment by petermattis** [comment]
https://github.com/cockroachdb/cockroach/compare/master...petermattis:pmattis/benchmark-track-choices?expand=1 is @tschottdorf's `tblgen` converted to a benchmark.

**Comment by tschottdorf** [comment]
I did more poking around today using the benchmark and a more practical approach (running tblgen and adding instrumentation). See the bulk-load branch.
Essentially, I looked at the worst "removable" offender (not caring about consistency) on a fresh single-node and disabled it.

TL;DR: with many hacks, down from 172µs to 66µs, around a factor of 4 away from Postgres (running on my OSX in Docker).

```
name              base.10s      manyhacks.10s  postgres.10s
TrackChoices1280  172µs  ±  0%  66µs  ±  0%    16µs  ±  0%
```

Hacks added:

* not checking tsCache:  171532ns -> 146628ns (-15%). This was expensive because we hit the interval tree for each request in the (~1280 element) batch. The check was also likely breaking all of the previously inserted "wide spans" since the insert load is random (i.e. not sorted).
* not populating tsCache: 147µs -> 91µs (-~40%). Ditto, but this time writing all the time. Pretty bad and huge savings here.
* force all writes to be blind puts (they weren't previously, but we were doing all of the work - the order of entries is random, so we find a huge run without duplicates, but then the iterator finds a key right away and nothing actually writes blindly): ~91µs to 66µs, shaving off another close to 30%.

Fully instrumented trace of the last commit. With most of the non-storage distractions gone, we see a sizable chunk in Raft and the actual meat of the write to the batch.

```
2016/05/10 20:15:44.886244  0.006944        node 1
20:15:44.886246      .     3        ... node 1
20:15:44.886253      .     6        ... read has no clock uncertainty
20:15:44.886334      .    81        ... executing 1282 requests
20:15:44.886543      .   209        ... read-write path
20:15:44.886544      .     1        ... command queue
20:15:44.886963      .   420        ... left command queue
20:15:44.886966      .     3        ... request leader lease (attempt #1)
20:15:44.887265      .   298        ... prep for ts cache
20:15:44.887266      .     1        ... applied ts cache
20:15:44.887268      .     2        ... start marshal
20:15:44.887772      .   504        ... end marshal
20:15:44.887835      .    63        ... begin prop
20:15:44.887852      .    17        ... done prop
20:15:44.887855      .     3        ... proposed to Raft
20:15:44.888949      .  1094        ... applying batch // this is essentially upstream-Raft and on hiccups, we see ~5000; not sure what's up there
20:15:44.889015      .    67        ... checked aborted txn
20:15:44.889197      .   181        ... checked leadership
20:15:44.889199      .     2        ... executing as 1PC txn
20:15:44.889644      .   445        ... 1280 blind puts
20:15:44.892245      .  2601        ... executeBatch returns
20:15:44.892308      .    63        ... executed batch
20:15:44.892360      .    52        ... prep for commit
20:15:44.893101      .   741        ... committed
20:15:44.893154      .    53        ... processed async intents
20:15:44.893155      .     1        ... applied batch
20:15:44.893173      .    18        ... response obtained
20:15:44.893174      .     1        ... endCmds begins
20:15:44.893175      .     1        ... begin update tsCache
20:15:44.893176      .     1        ... end update tsCache
20:15:44.893177      .     2        ... removed from cmdQ
20:15:44.893178      .     1        ... endCmds ends
20:15:44.893180      .     1        ... Send returns
```

**Comment by petermattis** [comment]
I'm starting to take a look at this and can confirm the above bottlenecks in the timestamp cache when using the `TrackChoices` benchmarks. The attached cpu profile is from running:

`go test -count 1 -v -run - -bench 'TrackChoices1000_Cockroach' -benchmem -benchtime 10s -cpuprofile cpu.prof`

[prof.pdf](https://github.com/cockroachdb/cockroach/files/261987/prof.pdf)

Notice the large amount of time in `TimestampCache.Add` and `applyTimestampCache`. There is likely some fruit to be plucked in not using an `interval.Tree` for these caches as we don't have overlapping intervals in them.

Note that in this benchmark we don't have any actual overlap of timestamps. Out of 6.55s in `TimestampCache.Add`, we see the following per-line cpu consumption:

```
         .          .    160:           // subtract from the range/ranges that need to be added to cache.
      20ms      1.48s    161:           for _, entry := range tcache.GetOverlaps(r.Start, r.End) {
         .          .    162:                   cv := entry.Value.(*cacheValue)
```

```
         .          .    188:                           case sCmp > 0 && eCmp < 0:
         .          .    189:                                   // Old contains new; split up old into two.
         .          .    190:                                   //
         .          .    191:                                   // New:     ----
         .          .    192:                                   // Old: ------------
         .          .    193:                                   //
         .          .    194:                                   // Old: ----    ----
         .          .    195:                                   oldEnd := key.End
         .          .    196:                                   key.End = r.Start
         .          .    197:
         .       10ms    198:                                   key := tcache.MakeKey(r.End, oldEnd)
         .      120ms    199:                                   newEntry := makeCacheEntry(key, *cv)
         .      750ms    200:                                   tcache.AddEntryAfter(newEntry, entry)
```

```
         .          .    266:           }
         .      4.12s    267:           addRange(r)
         .          .    268:   }
```

So we're spending 85% of the time in `TimestampCache.Add` checking for overlaps and adding a range to the cache.

**Comment by nvanbenschoten** [comment]
Given that we no longer have overlaps in the `TimestampCache`, I think we should investigate if replacing the interval tree with a range tree would be possible. It seems like each node in a range tree would map cleanly to an interval endpoint, and we would just need to store in the node if the range to the right is empty or contains a `cacheValue`. The RangeTree would then expose an interface somewhat like [this](https://github.com/Workiva/go-datastructures/blob/master/rangetree/interface.go#L49), which seems to support all operations we currently use. I don't have any firm theory to back this up yet, but I'd imagine not having to handle overlaps at all would allow for an improved algorithmic complexity in favor of range trees for most basic operations.

Additionally, range trees can be implemented fairly straightforwardly using a SkipList (see [go-datastructures](https://github.com/Workiva/go-datastructures/blob/master/rangetree/skiplist/skiplist.go)), so if we did want to explore a concurrent data-structure (#4768), that might be the way to go. Based off about a minute of Googling, it looks like concurrent skiplists are much easier to implement in a lock-free manner using available atomic operations than RBTrees, which may require transactional memory.

**Comment by petermattis** [comment]
Both of these ideas are worth exploring, but I'd like to finish off #6413 first. I think I see a way to apply the "covering span" optimization I used on the `CommandQueue` to the `TimestampCache`. While the skip-list and RangeTree approaches might make the `TimestampCache` 10-20% faster, performing several orders of magnitude fewer operations will be a much more significant win. Of course, we very well might end up implementing all of these ideas.

**Comment by petermattis** [comment]
The `TimestampCache` needs to be able to answer the query: what is the maximum read/write timestamp that has been seen for a span that is greater than a specified timestamp (the timestamp of the request being executed). Currently, `TimestampCache.GetMax{Read,Write}` answer the slightly more general query: what is the maximum read/write timestamp for a span. If we include the timestamp of the request we're querying the timestamp cache for, I think we can change `TimestampCache` to delay interval tree manipulations until a conflict occurs. Instead of adding all of the spans involved in a request when the request completes, we compute the read/write spans and maintain a tree mapping from `timestamp` to read/write spans. When we query the timestamp cache before executing a request, we first "expand" any previously executed requests that have a newer timestamp than the request being executed. In the common case where the request being executed has a timestamp greater than any previously executed request, no requests would be expanded. 

@tschottdorf, @nvanbenschoten Does the above look right to you?

**Comment by vivekmenezes** [comment]
This sounds like the, look at only spans from requests that conflict with a particular request on the time dimension. Expand the conflicting requests on demand to account for the worst case scenario where the node receives a deluge of requests from an earlier timestamp.

I think the above approach could benefit from not using a tree for the time dimension but a slice of intervals allowing quick update and traversal of all spans above a certain timestamp. We will still need the tree of spans for the worst case performance case.

**Comment by petermattis** [comment]
@vivekmenezes Yes, we will still need the tree of spans for the worse case scenario. I'm not clear what your other suggestion is, though. How would you avoid having a tree keyed by timestamp?

**Comment by petermattis** [comment]
@vivekmenezes Take a look at #6413 for a concrete implementation of my suggestion.

**Comment by vivekmenezes** [comment]
I've not given this much thought and I'm sure I'm missing something, but here goes:

Keep track of say the last 10 seconds of requests (timestamp cache by creating a slice of intervals over the 10 seconds say at a 5 millisecond granularity (2000 entries). As time marches forward you move around the slice as if it were a wheel of time slots. Each slot has a boolean denoting whether all the entries in the slot are in the tree of spans. Any request that reads data is placed in its time slot based on its timestamp (constant time), if its slot is also marked as in the tree of spans, it needs to be added to the tree (additional log n time).  When looking up requests in the wheel, you need only consider all requests in your slot and all slots ahead of you, this involves traversing the wheel from the latest slot to the last slot added, and then your slot forward to the last slot added (essentially adding stuff around the stuff already added), adding all entries in slots not marked as being in the tree, and performing a tree lookup at the end.

Such a schema makes sense if we're mostly deleting slots that have the boolean unset.

**Comment by petermattis** [comment]
Got it.

In my experience, it is better to use a tree structure (e.g. the btree in #6413) to deal with time-based expiration than a slot approach as you've suggested. That said, I'm willing to be proven wrong. I doubt this will be worth experimenting with though, as the approach in #6413 is going to capture almost all of the benefit (avoiding manipulation of the interval tree) and doesn't have any significant overhead that the slot approach will eliminate.

**Comment by spencerkimball** [comment]
Much has been done here. My timings are now as follows:

**CockroachDB**
real	0m0.642s
user	0m0.039s
sys	0m0.021s

**Postgres**
real	0m0.179s
user	0m0.017s
sys	0m0.011s

I'm closing this issue for the time being. Other issues are tracking the more advanced ingest pathways.

**Comment by mani-minmini** [comment]
czxczxczxc
